File: casper/src/rust/blocks/proposer/proposer.rs
// See casper/src/main/scala/coop/rchain/casper/blocks/proposer/Proposer.scala

use std::sync::{Arc, Mutex};
use tracing;

use block_storage::rust::{
    deploy::key_value_deploy_storage::KeyValueDeployStorage,
    key_value_block_store::KeyValueBlockStore,
};
use comm::rust::{
    rp::{connect::ConnectionsCell, rp_conf::RPConf},
    transport::transport_layer::TransportLayer,
};
use crypto::rust::private_key::PrivateKey;
use models::rust::casper::pretty_printer::PrettyPrinter;
use models::rust::casper::protocol::casper_message::BlockMessage;
use shared::rust::shared::f1r3fly_events::F1r3flyEvents;

use crate::rust::{
    block_status::{BlockError, InvalidBlock},
    blocks::proposer::{
        block_creator,
        propose_result::{
            BlockCreatorResult, CheckProposeConstraintsFailure, CheckProposeConstraintsResult,
            ProposeFailure, ProposeResult,
        },
    },
    casper::{Casper, CasperSnapshot},
    engine::block_retriever::BlockRetriever,
    errors::CasperError,
    last_finalized_height_constraint_checker,
    multi_parent_casper_impl::{self},
    synchrony_constraint_checker,
    util::rholang::runtime_manager::RuntimeManager,
    validator_identity::ValidatorIdentity,
    ValidBlockProcessing,
};

use super::propose_result::ProposeStatus;

pub struct ProposeReturnType {
    pub propose_result: ProposeResult,
    pub propose_result_to_send: ProposerResult,
    pub block_message_opt: Option<BlockMessage>,
}

// Traits for dependency injection and testing
#[allow(async_fn_in_trait)]
pub trait CasperSnapshotProvider {
    async fn get_casper_snapshot(
        &self,
        casper: Arc<dyn Casper + Send + Sync + 'static>,
    ) -> Result<CasperSnapshot, CasperError>;
}

pub trait ActiveValidatorChecker {
    fn check_active_validator(
        &self,
        casper_snapshot: &CasperSnapshot,
        validator_identity: &ValidatorIdentity,
    ) -> CheckProposeConstraintsResult;
}

#[allow(async_fn_in_trait)]
pub trait StakeChecker {
    async fn check_enough_base_stake(
        &self,
        casper_snapshot: &CasperSnapshot,
    ) -> Result<CheckProposeConstraintsResult, CasperError>;
}

#[allow(async_fn_in_trait)]
pub trait HeightChecker {
    async fn check_finalized_height(
        &self,
        casper_snapshot: &CasperSnapshot,
    ) -> Result<CheckProposeConstraintsResult, CasperError>;
}

#[allow(async_fn_in_trait)]
pub trait BlockCreator {
    async fn create_block(
        &mut self,
        casper_snapshot: &CasperSnapshot,
        validator_identity: &ValidatorIdentity,
        dummy_deploy_opt: Option<(PrivateKey, String)>,
        allow_empty_blocks: bool,
    ) -> Result<BlockCreatorResult, CasperError>;
}

#[allow(async_fn_in_trait)]
pub trait BlockValidator {
    async fn validate_block(
        &self,
        casper: Arc<dyn Casper + Send + Sync + 'static>,
        casper_snapshot: &mut CasperSnapshot,
        block: &BlockMessage,
    ) -> Result<ValidBlockProcessing, CasperError>;
}

#[allow(async_fn_in_trait)]
pub trait ProposeEffectHandler {
    async fn handle_propose_effect(
        &mut self,
        casper: Arc<dyn Casper + Send + Sync + 'static>,
        block: &BlockMessage,
    ) -> Result<(), CasperError>;

    /// Publish BlockCreated event immediately after block creation (before validation).
    fn publish_block_created(&self, block: &BlockMessage) -> Result<(), CasperError>;
}

pub enum ProposerResult {
    Empty,
    Success(ProposeStatus, BlockMessage),
    Failure(ProposeStatus, i32),
    Started(i32),
}

impl ProposerResult {
    pub fn empty() -> Self {
        Self::Empty
    }

    pub fn success(status: ProposeStatus, block: BlockMessage) -> Self {
        Self::Success(status, block)
    }

    pub fn failure(status: ProposeStatus, seq_number: i32) -> Self {
        Self::Failure(status, seq_number)
    }

    pub fn started(seq_number: i32) -> Self {
        Self::Started(seq_number)
    }
}

pub struct Proposer<C, A, S, H, BC, BV, E>
where
    C: CasperSnapshotProvider,
    A: ActiveValidatorChecker,
    S: StakeChecker,
    H: HeightChecker,
    BC: BlockCreator,
    BV: BlockValidator,
    E: ProposeEffectHandler,
{
    pub validator: Arc<ValidatorIdentity>,
    pub dummy_deploy_opt: Option<(PrivateKey, String)>,
    pub casper_snapshot_provider: C,
    pub active_validator_checker: A,
    pub stake_checker: S,
    pub height_checker: H,
    pub block_creator: BC,
    pub block_validator: BV,
    pub propose_effect_handler: E,
    /// When true, allows creating blocks with only system deploys (no user deploys).
    /// This is required for heartbeat to create empty blocks for liveness.
    pub allow_empty_blocks: bool,
}

impl<C, A, S, H, BC, BV, E> Proposer<C, A, S, H, BC, BV, E>
where
    C: CasperSnapshotProvider,
    A: ActiveValidatorChecker,
    S: StakeChecker,
    H: HeightChecker,
    BC: BlockCreator,
    BV: BlockValidator,
    E: ProposeEffectHandler,
{
    pub fn new(
        validator: Arc<ValidatorIdentity>,
        dummy_deploy_opt: Option<(PrivateKey, String)>,
        casper_snapshot_provider: C,
        active_validator_checker: A,
        stake_checker: S,
        height_checker: H,
        block_creator: BC,
        block_validator: BV,
        propose_effect_handler: E,
        allow_empty_blocks: bool,
    ) -> Self {
        Self {
            validator,
            dummy_deploy_opt,
            casper_snapshot_provider,
            active_validator_checker,
            stake_checker,
            height_checker,
            block_creator,
            block_validator,
            propose_effect_handler,
            allow_empty_blocks,
        }
    }

    // This is the whole logic of propose
    async fn do_propose(
        &mut self,
        casper_snapshot: &mut CasperSnapshot,
        casper: Arc<dyn Casper + Send + Sync + 'static>,
        allow_empty_blocks_for_request: bool,
    ) -> Result<(ProposeResult, Option<BlockMessage>), CasperError> {
        // check if node is allowed to propose a block
        let constraint_check = self.check_propose_constraints(casper_snapshot).await?;

        match constraint_check {
            CheckProposeConstraintsResult::Failure(failure) => Ok((
                ProposeResult::failure(ProposeFailure::CheckConstraintsFailure(failure)),
                None,
            )),
            CheckProposeConstraintsResult::Success => {
                let block_result = self
                    .block_creator
                    .create_block(
                        casper_snapshot,
                        &self.validator,
                        self.dummy_deploy_opt.clone(),
                        allow_empty_blocks_for_request,
                    )
                    .await?;

                match block_result {
                    BlockCreatorResult::NoNewDeploys => {
                        Ok((ProposeResult::failure(ProposeFailure::NoNewDeploys), None))
                    }
                    BlockCreatorResult::Created(block, pre_state_hash, post_state_hash) => {
                        // Publish BlockCreated event immediately after block is created (before validation)
                        self.propose_effect_handler.publish_block_created(&block)?;

                        let validation_result = casper
                            .validate_self_created(
                                &block,
                                casper_snapshot,
                                pre_state_hash,
                                post_state_hash,
                            )
                            .await?;

                        match validation_result {
                            ValidBlockProcessing::Right(valid_status) => {
                                self.propose_effect_handler
                                    .handle_propose_effect(casper, &block)
                                    .await?;
                                Ok((ProposeResult::success(valid_status), Some(block)))
                            }
                            ValidBlockProcessing::Left(invalid_reason) => {
                                // Some self-validation failures are recoverable races in fast, multi-parent
                                // proposing: parent selection can become stale, and safety checks can reject
                                // the candidate by the time validation runs.
                                if matches!(
                                    invalid_reason,
                                    BlockError::Invalid(InvalidBlock::InvalidParents)
                                        | BlockError::Invalid(InvalidBlock::InvalidFollows)
                                        | BlockError::Invalid(
                                            InvalidBlock::JustificationRegression
                                        )
                                        | BlockError::Invalid(InvalidBlock::InvalidBondsCache)
                                        | BlockError::Invalid(InvalidBlock::InvalidRepeatDeploy)
                                        | BlockError::Invalid(InvalidBlock::NeglectedInvalidBlock)
                                ) {
                                    let recoverable_reason = match &invalid_reason {
                                        BlockError::Invalid(InvalidBlock::InvalidParents) => {
                                            "invalid_parents"
                                        }
                                        BlockError::Invalid(InvalidBlock::InvalidFollows) => {
                                            "invalid_follows"
                                        }
                                        BlockError::Invalid(InvalidBlock::InvalidBondsCache) => {
                                            "invalid_bonds_cache"
                                        }
                                        BlockError::Invalid(InvalidBlock::InvalidRepeatDeploy) => {
                                            "invalid_repeat_deploy"
                                        }
                                        BlockError::Invalid(
                                            InvalidBlock::JustificationRegression,
                                        ) => "justification_regression",
                                        BlockError::Invalid(
                                            InvalidBlock::NeglectedInvalidBlock,
                                        ) => "neglected_invalid_block",
                                        _ => "other",
                                    };
                                    metrics::counter!(
                                        "propose_recoverable_self_validation_failures_total",
                                        "source" => "casper_proposer",
                                        "reason" => recoverable_reason
                                    )
                                    .increment(1);
                                    tracing::info!(
                                        "Block validation failed with {:?} - \
                                         proposal conditions no longer met, skipping propose",
                                        invalid_reason
                                    );
                                    return Ok((
                                        ProposeResult::failure(ProposeFailure::InternalDeployError),
                                        None,
                                    ));
                                }

                                // Other validation failures are unexpected and should error
                                Err(CasperError::RuntimeError(format!(
                                    "Validation of self created block failed with reason: {:?}, cancelling propose.",
                                    invalid_reason
                                )))
                            }
                        }
                    }
                }
            }
        }
    }

    // Check if proposer can issue a block
    pub async fn check_propose_constraints(
        &mut self,
        casper_snapshot: &CasperSnapshot,
    ) -> Result<CheckProposeConstraintsResult, CasperError> {
        match self
            .active_validator_checker
            .check_active_validator(casper_snapshot, &self.validator)
        {
            CheckProposeConstraintsResult::Failure(CheckProposeConstraintsFailure::NotBonded) => {
                Ok(CheckProposeConstraintsResult::not_bonded())
            }
            _ => {
                // Run both async checks in parallel
                let (stake_result, height_result) = tokio::join!(
                    self.stake_checker.check_enough_base_stake(casper_snapshot),
                    self.height_checker.check_finalized_height(casper_snapshot)
                );

                // Handle any errors from the async calls
                let stake_check = stake_result?;
                let height_check = height_result?;

                // Pick some result that is not Success, or return Success
                let results = vec![stake_check, height_check];
                for result in results {
                    match result {
                        CheckProposeConstraintsResult::Success => continue,
                        failure => return Ok(failure),
                    }
                }

                Ok(CheckProposeConstraintsResult::success())
            }
        }
    }

    pub async fn propose(
        &mut self,
        casper: Arc<dyn Casper + Send + Sync + 'static>,
        is_async: bool,
    ) -> Result<ProposeReturnType, CasperError> {
        // Using tracing events instead of spans for async context
        // Span[F].traceI("do-propose") equivalent from Scala
        tracing::info!(target: "f1r3fly.casper.proposer", "do-propose-started");
        tracing::debug!(target: "f1r3fly.casper.proposer", "started-do-propose");

        fn get_validator_next_seq_number(
            casper_snapshot: &CasperSnapshot,
            validator_public_key: &[u8],
        ) -> i32 {
            casper_snapshot
                .max_seq_nums
                .get(validator_public_key)
                .map(|seq| *seq + 1)
                .unwrap_or(1) as i32
        }

        let start_time = std::time::Instant::now();

        // get snapshot to serve as a base for propose
        let snapshot_start = std::time::Instant::now();
        let mut casper_snapshot = self
            .casper_snapshot_provider
            .get_casper_snapshot(casper.clone())
            .await?;
        let snapshot_ms = snapshot_start.elapsed().as_millis();

        let elapsed = start_time.elapsed();
        tracing::info!("getCasperSnapshot [{}ms]", elapsed.as_millis());

        let self_seq = casper_snapshot
            .max_seq_nums
            .get(&self.validator.public_key.bytes)
            .map(|seq| *seq as i64)
            .unwrap_or(0);
        let observed_max_seq = casper_snapshot
            .max_seq_nums
            .iter()
            .map(|entry| *entry.value())
            .max()
            .unwrap_or(0) as i64;
        let (block_lag, seq_lag) = match casper_snapshot
            .dag
            .lookup_unsafe(&casper_snapshot.last_finalized_block)
        {
            Ok(lfb_meta) => (
                casper_snapshot
                    .max_block_num
                    .saturating_sub(lfb_meta.block_number),
                observed_max_seq.saturating_sub(lfb_meta.sequence_number as i64),
            ),
            Err(_) => (0, 0),
        };
        let finality_lag = std::cmp::max(block_lag, seq_lag);
        let allow_empty_for_recovery = finality_lag > 20;
        if allow_empty_for_recovery && !is_async {
            tracing::info!(
                "Enabling empty-block propose in sync recovery mode due to finality lag (lag={}, block_lag={}, seq_lag={}, self_seq={}, observed_max_seq={})",
                finality_lag,
                block_lag,
                seq_lag,
                self_seq,
                observed_max_seq
            );
        }

        let allow_empty_blocks_for_request =
            (self.allow_empty_blocks && is_async) || allow_empty_for_recovery;

        let (result, propose_core_ms) = if is_async {
            // Empty blocks are reserved for heartbeat/liveness-driven proposes.
            // Synchronous/manual propose calls should fail fast when no new deploys exist.
            // propose
            let propose_start = std::time::Instant::now();
            let (propose_result, block_opt) = self
                .do_propose(
                    &mut casper_snapshot,
                    casper.clone(),
                    allow_empty_blocks_for_request,
                )
                .await?;
            let propose_core_ms = propose_start.elapsed().as_millis();

            let propose_result_to_send = match &block_opt {
                Some(block) => {
                    ProposerResult::success(propose_result.propose_status.clone(), block.clone())
                }
                None => {
                    let next_seq = get_validator_next_seq_number(
                        &casper_snapshot,
                        &self.validator.public_key.bytes,
                    );
                    ProposerResult::failure(propose_result.propose_status.clone(), next_seq)
                }
            };

            (
                ProposeReturnType {
                    propose_result,
                    propose_result_to_send,
                    block_message_opt: block_opt,
                },
                propose_core_ms,
            )
        } else {
            // Empty blocks are reserved for heartbeat/liveness-driven proposes.
            // Synchronous/manual propose calls should fail fast when no new deploys exist.
            // propose
            let propose_start = std::time::Instant::now();
            let (propose_result, block_opt) = self
                .do_propose(&mut casper_snapshot, casper, allow_empty_blocks_for_request)
                .await?;
            let propose_core_ms = propose_start.elapsed().as_millis();

            let propose_result_to_send = match &block_opt {
                None => {
                    let seq_number = get_validator_next_seq_number(
                        &casper_snapshot,
                        &self.validator.public_key.bytes,
                    );
                    ProposerResult::failure(propose_result.propose_status.clone(), seq_number)
                }
                Some(block) => {
                    ProposerResult::success(propose_result.propose_status.clone(), block.clone())
                }
            };

            (
                ProposeReturnType {
                    propose_result,
                    propose_result_to_send,
                    block_message_opt: block_opt,
                },
                propose_core_ms,
            )
        };

        let total_ms = start_time.elapsed().as_millis();
        tracing::info!(
            target: "f1r3fly.propose.timing",
            "Propose timing: mode={}, snapshot_ms={}, propose_core_ms={}, total_ms={}",
            if is_async { "async" } else { "sync" },
            snapshot_ms,
            propose_core_ms,
            total_ms
        );

        tracing::debug!(target: "f1r3fly.casper.proposer", "finished-do-propose");
        tracing::info!(target: "f1r3fly.casper.proposer", "do-propose-finished");
        Ok(result)
    }
}

pub type ProductionProposer<T> = Proposer<
    ProductionCasperSnapshotProvider,
    ProductionActiveValidatorChecker,
    ProductionStakeChecker,
    ProductionHeightChecker,
    ProductionBlockCreator,
    ProductionBlockValidator,
    ProductionProposeEffectHandler<T>,
>;

pub fn new_proposer<T: TransportLayer + Send + Sync + 'static>(
    validator: ValidatorIdentity,
    dummy_deploy_opt: Option<(PrivateKey, String)>,
    runtime_manager: RuntimeManager,
    block_store: KeyValueBlockStore,
    deploy_storage: Arc<Mutex<KeyValueDeployStorage>>,
    block_retriever: BlockRetriever<T>,
    transport: Arc<T>,
    connections_cell: ConnectionsCell,
    conf: RPConf,
    event_publisher: F1r3flyEvents,
    allow_empty_blocks: bool,
) -> ProductionProposer<T> {
    let validator_arc = Arc::new(validator);

    Proposer::new(
        validator_arc.clone(),
        dummy_deploy_opt,
        ProductionCasperSnapshotProvider,
        ProductionActiveValidatorChecker,
        ProductionStakeChecker::new(
            runtime_manager.clone(),
            block_store.clone(),
            validator_arc.clone(),
        ),
        ProductionHeightChecker::new(validator_arc),
        ProductionBlockCreator::new(deploy_storage, runtime_manager.clone(), block_store.clone()),
        ProductionBlockValidator,
        ProductionProposeEffectHandler::new(
            block_store,
            block_retriever,
            transport,
            connections_cell,
            conf,
            event_publisher,
        ),
        allow_empty_blocks,
    )
}

pub struct ProductionCasperSnapshotProvider;
impl CasperSnapshotProvider for ProductionCasperSnapshotProvider {
    async fn get_casper_snapshot(
        &self,
        casper: Arc<dyn Casper + Send + Sync + 'static>,
    ) -> Result<CasperSnapshot, CasperError> {
        casper.get_snapshot().await
    }
}

pub struct ProductionActiveValidatorChecker;
impl ActiveValidatorChecker for ProductionActiveValidatorChecker {
    fn check_active_validator(
        &self,
        casper_snapshot: &CasperSnapshot,
        validator_identity: &ValidatorIdentity,
    ) -> CheckProposeConstraintsResult {
        if casper_snapshot
            .on_chain_state
            .active_validators
            .contains(&validator_identity.public_key.bytes)
        {
            CheckProposeConstraintsResult::success()
        } else {
            CheckProposeConstraintsResult::not_bonded()
        }
    }
}

pub struct ProductionStakeChecker {
    runtime_manager: RuntimeManager,
    block_store: KeyValueBlockStore,
    validator: Arc<ValidatorIdentity>,
}

impl ProductionStakeChecker {
    pub fn new(
        runtime_manager: RuntimeManager,
        block_store: KeyValueBlockStore,
        validator: Arc<ValidatorIdentity>,
    ) -> Self {
        Self {
            runtime_manager,
            block_store,
            validator,
        }
    }
}

impl StakeChecker for ProductionStakeChecker {
    async fn check_enough_base_stake(
        &self,
        casper_snapshot: &CasperSnapshot,
    ) -> Result<CheckProposeConstraintsResult, CasperError> {
        synchrony_constraint_checker::check(
            casper_snapshot,
            &self.runtime_manager,
            &self.block_store,
            &self.validator,
        )
        .await
    }
}

pub struct ProductionHeightChecker {
    validator: Arc<ValidatorIdentity>,
}

impl ProductionHeightChecker {
    pub fn new(validator: Arc<ValidatorIdentity>) -> Self {
        Self { validator }
    }
}

impl HeightChecker for ProductionHeightChecker {
    async fn check_finalized_height(
        &self,
        casper_snapshot: &CasperSnapshot,
    ) -> Result<CheckProposeConstraintsResult, CasperError> {
        last_finalized_height_constraint_checker::check(casper_snapshot, &self.validator)
    }
}

pub struct ProductionBlockCreator {
    deploy_storage: Arc<Mutex<KeyValueDeployStorage>>,
    runtime_manager: RuntimeManager,
    block_store: KeyValueBlockStore,
}

impl ProductionBlockCreator {
    pub fn new(
        deploy_storage: Arc<Mutex<KeyValueDeployStorage>>,
        runtime_manager: RuntimeManager,
        block_store: KeyValueBlockStore,
    ) -> Self {
        Self {
            deploy_storage,
            runtime_manager,
            block_store,
        }
    }
}

impl BlockCreator for ProductionBlockCreator {
    async fn create_block(
        &mut self,
        casper_snapshot: &CasperSnapshot,
        validator_identity: &ValidatorIdentity,
        dummy_deploy_opt: Option<(PrivateKey, String)>,
        allow_empty_blocks: bool,
    ) -> Result<BlockCreatorResult, CasperError> {
        block_creator::create(
            casper_snapshot,
            validator_identity,
            dummy_deploy_opt,
            self.deploy_storage.clone(),
            &mut self.runtime_manager,
            &mut self.block_store,
            allow_empty_blocks,
        )
        .await
    }
}

pub struct ProductionBlockValidator;
impl BlockValidator for ProductionBlockValidator {
    async fn validate_block(
        &self,
        casper: Arc<dyn Casper + Send + Sync + 'static>,
        casper_snapshot: &mut CasperSnapshot,
        block: &BlockMessage,
    ) -> Result<ValidBlockProcessing, CasperError> {
        casper.validate(block, casper_snapshot).await
    }
}

pub struct ProductionProposeEffectHandler<T: TransportLayer + Send + Sync> {
    block_store: KeyValueBlockStore,
    block_retriever: BlockRetriever<T>,
    transport: Arc<T>,
    connections_cell: ConnectionsCell,
    conf: RPConf,
    event_publisher: F1r3flyEvents,
}

impl<T: TransportLayer + Send + Sync> ProductionProposeEffectHandler<T> {
    pub fn new(
        block_store: KeyValueBlockStore,
        block_retriever: BlockRetriever<T>,
        transport: Arc<T>,
        connections_cell: ConnectionsCell,
        conf: RPConf,
        event_publisher: F1r3flyEvents,
    ) -> Self {
        Self {
            block_store,
            block_retriever,
            transport,
            connections_cell,
            conf,
            event_publisher,
        }
    }
}

impl<T: TransportLayer + Send + Sync + 'static> ProposeEffectHandler
    for ProductionProposeEffectHandler<T>
{
    async fn handle_propose_effect(
        &mut self,
        casper: Arc<dyn Casper + Send + Sync + 'static>,
        block: &BlockMessage,
    ) -> Result<(), CasperError> {
        // store block
        self.block_store.put_block_message(block)?;

        // save changes to Casper (publishes BlockAdded and BlockFinalised events)
        casper.handle_valid_block(block).await?;

        // inform block retriever about block
        self.block_retriever
            .ack_in_casper(block.block_hash.clone())
            .await?;

        // Broadcast hash to peers on a best-effort basis, but do not let network fan-out tail
        // block local propose completion latency.
        let transport = Arc::clone(&self.transport);
        let connections_cell = self.connections_cell.clone();
        let conf = self.conf.clone();
        let block_hash = block.block_hash.clone();
        let sender = block.sender.clone();
        tokio::spawn(async move {
            if let Err(err) = transport
                .send_block_hash(&connections_cell, &conf, &block_hash, &sender)
                .await
            {
                tracing::warn!(
                    "Failed to broadcast block hash {} to some peers: {}",
                    PrettyPrinter::build_string_bytes(&block_hash),
                    err
                );
            }
        });

        Ok(())
    }

    fn publish_block_created(&self, block: &BlockMessage) -> Result<(), CasperError> {
        // Publish BlockCreated event
        self.event_publisher
            .publish(multi_parent_casper_impl::created_event(block))
            .map_err(|e| CasperError::RuntimeError(e.to_string()))
    }
}

File: casper/src/main/scala/coop/rchain/casper/blocks/proposer/Proposer.scala
package coop.rchain.casper.blocks.proposer

import cats.effect.Concurrent
import cats.effect.concurrent.Deferred
import cats.syntax.all._
import com.google.protobuf.ByteString
import coop.rchain.blockstorage.BlockStore
import coop.rchain.blockstorage.dag.BlockDagStorage
import coop.rchain.blockstorage.deploy.DeployStorage
import coop.rchain.casper.engine.BlockRetriever
import coop.rchain.casper.protocol.BlockMessage
import coop.rchain.casper.syntax._
import coop.rchain.casper.util.comm.CommUtil
import coop.rchain.casper.util.rholang.RuntimeManager
import coop.rchain.casper._
import coop.rchain.crypto.PrivateKey
import coop.rchain.metrics.Metrics.Source
import coop.rchain.metrics.implicits._
import coop.rchain.metrics.{Metrics, Span}
import coop.rchain.shared.{EventPublisher, Log, Stopwatch, Time}
import fs2.Stream

sealed abstract class ProposerResult
object ProposerEmpty                                                         extends ProposerResult
final case class ProposerSuccess(status: ProposeStatus, block: BlockMessage) extends ProposerResult
final case class ProposerFailure(status: ProposeStatus, seqNumber: Int)      extends ProposerResult
final case class ProposerStarted(seqNumber: Int)                             extends ProposerResult

object ProposerResult {
  def empty: ProposerResult = ProposerEmpty
  def success(status: ProposeStatus, block: BlockMessage): ProposerResult =
    ProposerSuccess(status, block)
  def failure(status: ProposeStatus, seqNumber: Int): ProposerResult =
    ProposerFailure(status, seqNumber)
  def started(seqNumber: Int): ProposerResult = ProposerStarted(seqNumber)
}

class Proposer[F[_]: Concurrent: Log: Span: EventPublisher](
    // base state on top of which block will be created
    getCasperSnapshot: Casper[F] => F[CasperSnapshot[F]],
    // propose constraint checkers
    checkActiveValidator: (
        CasperSnapshot[F],
        ValidatorIdentity
    ) => CheckProposeConstraintsResult,
    checkEnoughBaseStake: (BlockMessage, CasperSnapshot[F]) => F[CheckProposeConstraintsResult],
    checkFinalizedHeight: (BlockMessage, CasperSnapshot[F]) => F[CheckProposeConstraintsResult],
    createBlock: (
        CasperSnapshot[F],
        ValidatorIdentity
    ) => F[BlockCreatorResult],
    validateBlock: (Casper[F], CasperSnapshot[F], BlockMessage) => F[ValidBlockProcessing],
    proposeEffect: (Casper[F], BlockMessage) => F[Unit],
    validator: ValidatorIdentity
) {

  implicit val RuntimeMetricsSource: Source = Metrics.Source(CasperMetricsSource, "proposer")
  // This is the whole logic of propose
  private def doPropose(
      s: CasperSnapshot[F],
      casper: Casper[F]
  ): F[(ProposeResult, Option[BlockMessage])] =
    Span[F].traceI("do-propose") {
      for {
        // TODO this genesis should not be here, but required for sync constraint code. Remove
        genesis <- casper.getApprovedBlock
        // check if node is allowed to propose a block
        chk <- checkProposeConstraints(genesis, s)
        r <- chk match {
              case v: CheckProposeConstraintsFailure =>
                (ProposeResult.failure(v), none[BlockMessage]).pure[F]
              case CheckProposeConstraintsSuccess =>
                for {
                  b <- createBlock(s, validator)
                  r <- b match {
                        case NoNewDeploys =>
                          (ProposeResult.failure(NoNewDeploys), none[BlockMessage]).pure[F]
                        case Created(b) =>
                          // Publish BlockCreated event immediately after block is created
                          EventPublisher[F].publish(MultiParentCasperImpl.createdEvent(b)) >>
                            validateBlock(casper, s, b).flatMap {
                              case Right(v) =>
                                proposeEffect(casper, b) >>
                                  (ProposeResult.success(v), b.some).pure[F]
                              case Left(v) =>
                                v match {
                                  // Transient conditions: DAG/state changed between snapshot
                                  // and block creation. These are expected in a concurrent
                                  // multi-validator network and will resolve on the next
                                  // heartbeat cycle with a fresh snapshot.
                                  // InvalidTimestamp: a new parent block arrived between block
                                  // creation and self-validation, making the timestamp stale
                                  // relative to the updated parent set.
                                  case InvalidBlock.InvalidParents |
                                      InvalidBlock.InvalidBondsCache |
                                      InvalidBlock.InvalidTransaction |
                                      InvalidBlock.InvalidRejectedDeploy |
                                      InvalidBlock.ContainsExpiredDeploy |
                                      InvalidBlock.ContainsTimeExpiredDeploy |
                                      InvalidBlock.InvalidTimestamp =>
                                    Log[F].error(
                                      s"Self-created block validation failed with transient reason: $v -- discarding block and will retry on next heartbeat"
                                    ) >>
                                      (
                                        ProposeResult.failure(InternalDeployError),
                                        none[BlockMessage]
                                      ).pure[F]

                                  // Structural errors: indicate a bug in block creation code.
                                  // These will never self-heal on retry, so crash to make
                                  // the problem immediately visible to the operator.
                                  case _ =>
                                    Concurrent[F].raiseError[(ProposeResult, Option[BlockMessage])](
                                      new Throwable(
                                        s"Self-created block validation failed with structural error: $v -- this indicates a bug in block creation"
                                      )
                                    )
                                }
                            }
                      }
                } yield r
            }
      } yield r
    }

  // Check if proposer can issue a block
  private def checkProposeConstraints(
      genesis: BlockMessage,
      s: CasperSnapshot[F]
  ): F[CheckProposeConstraintsResult] =
    checkActiveValidator(s, validator) match {
      case NotBonded => CheckProposeConstraintsResult.notBonded.pure[F]
      case _ =>
        val work = Stream(
          Stream.eval[F, CheckProposeConstraintsResult](checkEnoughBaseStake(genesis, s)),
          Stream.eval[F, CheckProposeConstraintsResult](checkFinalizedHeight(genesis, s))
        )
        work
          .parJoin(2)
          .compile
          .toList
          // pick some result that is not Success, or return Success
          .map(
            _.find(_ != CheckProposeConstraintsSuccess).getOrElse(CheckProposeConstraintsSuccess)
          )
    }

  def propose(
      c: Casper[F],
      isAsync: Boolean,
      proposeIdDef: Deferred[F, ProposerResult]
  ): F[(ProposeResult, Option[BlockMessage])] = {
    def getValidatorNextSeqNumber(cs: CasperSnapshot[F]): Int = {
      val valBytes = ByteString.copyFrom(validator.publicKey.bytes)
      cs.maxSeqNums.getOrElse(valBytes, 0) + 1
    }
    val work = for {
      // get snapshot to serve as a base for propose
      s <- Stopwatch.time(Log[F].info(_))(s"getCasperSnapshot")(getCasperSnapshot(c))
      result <- if (isAsync) for {
                 nextSeq <- getValidatorNextSeqNumber(s).pure[F]
                 _       <- proposeIdDef.complete(ProposerResult.started(nextSeq))

                 // propose
                 r <- doPropose(s, c)
               } yield r
               else
                 for {
                   // propose
                   r <- doPropose(s, c)

                   (result, blockHashOpt) = r
                   proposerResult = blockHashOpt.fold {
                     val seqNumber = getValidatorNextSeqNumber(s)
                     ProposerResult.failure(result.proposeStatus, seqNumber)
                   } { block =>
                     ProposerResult.success(result.proposeStatus, block)
                   }
                   _ <- proposeIdDef.complete(proposerResult)
                 } yield r

    } yield result

    work.handleErrorWith {
      case _: FinalizationInProgressException =>
        // Finalization is in progress -- the snapshot cannot be obtained right now.
        // This is a transient condition that resolves within seconds once finalization
        // completes. Skip this propose cycle; the heartbeat will trigger another attempt.
        Log[F].info(
          "Snapshot unavailable: finalization in progress, skipping propose cycle"
        ) >>
          proposeIdDef.complete(ProposerResult.empty).attempt.void >>
          (ProposeResult.failure(InternalDeployError), none[BlockMessage]).pure[F]
    }
  }
}

object Proposer {
  // format: off
  def apply[F[_]
    /* Execution */   : Concurrent: Time
    /* Casper */      : Estimator: SynchronyConstraintChecker: LastFinalizedHeightConstraintChecker
    /* Storage */     : BlockStore: BlockDagStorage: DeployStorage
    /* Diagnostics */ : Log: Span: Metrics: EventPublisher
    /* Comm */        : CommUtil: BlockRetriever
  ] // format: on
  (
      validatorIdentity: ValidatorIdentity,
      dummyDeployOpt: Option[(PrivateKey, String)] = None,
      allowEmptyBlocks: Boolean = false
  )(implicit runtimeManager: RuntimeManager[F]): Proposer[F] = {
    val getCasperSnapshotSnapshot = (c: Casper[F]) => c.getSnapshot

    val createBlock = (s: CasperSnapshot[F], validatorIdentity: ValidatorIdentity) =>
      BlockCreator.create(s, validatorIdentity, dummyDeployOpt, allowEmptyBlocks)

    val validateBlock = (casper: Casper[F], s: CasperSnapshot[F], b: BlockMessage) =>
      casper.validate(b, s)

    val checkValidatorIsActive = (s: CasperSnapshot[F], validator: ValidatorIdentity) =>
      if (s.onChainState.activeValidators.contains(ByteString.copyFrom(validator.publicKey.bytes)))
        CheckProposeConstraintsSuccess
      else
        NotBonded

    val checkEnoughBaseStake = (genesis: BlockMessage, s: CasperSnapshot[F]) =>
      SynchronyConstraintChecker[F].check(
        s,
        runtimeManager,
        genesis,
        validatorIdentity
      )

    val checkLastFinalizedHeightConstraint = (genesis: BlockMessage, s: CasperSnapshot[F]) =>
      LastFinalizedHeightConstraintChecker[F].check(
        s,
        genesis: BlockMessage,
        validatorIdentity
      )

    val proposeEffect = (c: Casper[F], b: BlockMessage) =>
      // store block
      BlockStore[F].put(b) >>
        // save changes to Casper (publishes BlockAdded and BlockFinalised)
        c.handleValidBlock(b) >>
        // inform block retriever about block
        BlockRetriever[F].ackInCasper(b.blockHash) >>
        // broadcast hash to peers
        CommUtil[F].sendBlockHash(b.blockHash, b.sender)

    new Proposer(
      getCasperSnapshotSnapshot,
      checkValidatorIsActive,
      checkEnoughBaseStake,
      checkLastFinalizedHeightConstraint,
      createBlock,
      validateBlock,
      proposeEffect,
      validatorIdentity
    )
  }
}
